File: code/algomatrix/Q2 Data Structures Correction(Please Ignore Others)/Group 6/HashTable.js
class HashTable {
    constructor(size = 10) {
        this.size = size;
        this.buckets = new Array(size).fill(null).map(() => []);
    }

    hash(key) {
        let hashValue = 0;
        for (let i = 0; i < key.length; i++) {
            hashValue += key.charCodeAt(i);
        }
        return hashValue % this.size;
    }

    set(key, value) {
        const index = this.hash(key);
        for (let kvp of this.buckets[index]) {
            if (kvp[0] === key) {
                kvp[1] = value; // Update existing key
                return;
            }
        }
        this.buckets[index].push([key, value]); // Add new key
    }

    get(key) {
        const index = this.hash(key);
        for (let kvp of this.buckets[index]) {
            if (kvp[0] === key) {
                return kvp[1];
            }
        }
        return undefined; // Key not found
    }

    delete(key) {
        const index = this.hash(key);
        this.buckets[index] = this.buckets[index].filter(kvp => kvp[0] !== key);
    }

    display() {
        for (let i = 0; i < this.size; i++) {
            console.log(`Bucket ${i}:`);
            for (let kvp of this.buckets[i]) {
                console.log(`  ${kvp[0]}: ${kvp[1]}`);
            }
        }
    }
}

File: code/algomatrix/Group 3/HashTable.java
import java.util.LinkedList;

class HashTable<K, V> {
    private static final int DEFAULT_CAPACITY = 10;
    private LinkedList<Entry<K, V>>[] buckets;
    private int size;

    private static class Entry<K, V> {
        K key;
        V value;

        Entry(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }

    public HashTable() {
        this(DEFAULT_CAPACITY);
    }

    public HashTable(int capacity) {
        this.buckets = new LinkedList[capacity];
        this.size = 0;
    }

    private int hash(K key) {
        return Math.abs(key.hashCode()) % buckets.length;
    }

    public void put(K key, V value) {
        int index = hash(key);
        if (buckets[index] == null) {
            buckets[index] = new LinkedList<>();
        }
        for (Entry<K, V> entry : buckets[index]) {
            if (entry.key.equals(key)) {
                entry.value = value; // Update existing key
                return;
            }
        }
        buckets[index].add(new Entry<>(key, value)); // Add new key
        size++;
    }

    public V get(K key) {
        int index = hash(key);
        if (buckets[index] != null) {
            for (Entry<K, V> entry : buckets[index]) {
                if (entry.key.equals(key)) {
                    return entry.value;
                }
            }
        }
        return null; // Key not found
    }

    public void remove(K key) {
        int index = hash(key);
        if (buckets[index] != null) {
            for (Entry<K, V> entry : buckets[index]) {
                if (entry.key.equals(key)) {
                    buckets[index].remove(entry);
                    size--;
                    return;
                }
            }
        }
    }

    public int size() {
        return size;
    }

    public void display() {
        for (int i = 0; i < buckets.length; i++) {
            if (buckets[i] != null) {
                System.out.print("Bucket " + i + ": ");
                for (Entry<K, V> entry : buckets[i]) {
                    System.out.print("(" + entry.key + ", " + entry.value + ") ");
                }
                System.out.println();
            }
        }
    }

    public static void main(String[] args) {
        HashTable<String, Integer> table = new HashTable<>();
        table.put("apple", 10);
        table.put("banana", 20);
        table.put("orange", 30);

        System.out.println("Value for key 'apple': " + table.get("apple")); // Output: 10
        System.out.println("Value for key 'banana': " + table.get("banana")); // Output: 20
        System.out.println("Value for key 'orange': " + table.get("orange")); // Output: 30

        table.remove("banana");
        table.display();
        // Output:
        // Bucket 0: (orange, 30)
        // Bucket 3: (apple, 10)
    }
}